snafu/benchmarks/dnsperf/dnsperf.py

```diff
@@ -0,0 +1,300 @@
+        self.config.parse_args()
+
+        if self.config.pod_id:
+            self.tmp_filepath += f"-{self.pod_id}"
+        if check_file(self.tmp_filepath):
+            self.logger.critical((f"Temporary file located at {self.tmp_filepath} " "already exists."))
+            return False
+        try:
+            tmp_file = open(self.tmp_filepath, "x")
+            tmp_file.close()
+        except Exception as e:
+            self.logger.critical(
+                f"Unable to create temporary file at {self.tmp_filepath}: {e}", exc_info=True
+            )
+            return False
+        else:
+            self.logger.info(f"Successfully create temporary file: {self.tmp_filepath}")
+        return True
+
+    def cleanup(self) -> bool:
+        """Clean up artifacts from the dnsperf benchmark."""
+        self.logger.info("Cleaning up dnsperf benchmark.")
+        try:
+            os.remove(self.tmp_filepath)
+        except Exception as e:
+            self.logger.critical(f"Unable to remove temporary file {self.tmp_filepath}: {e}", exc_info=True)
+            return False
+        self.logger.info(f"Successfully removed temp file {self.tmp_filepath}")
+        return True
+
+    def _one_trial(self, load=None) -> BenchmarkResult:
+        cmd = [
+            "dnsperf",
+            "-v",  # print latency information for each query
+            "-s",
+            self.config.address,
+            "-p",
+            self.config.port,
+            "-d",
+            self.config.query_filepath,
+            "-c",
+            self.config.clients,
+            "-l",
+            self.config.time_limit,
+            "-m",
+            self.config.transport_mode,
+        ]
+
+        if load:
+            cmd = [*cmd, "-Q", str(load)]
+
+        sample: ProcessSample = sample_process(
+            cmd, self.logger,
+        )
+
+        if not sample.success:
+            self.logger.critical(f"dnsperf failed to complete! Got results: {sample}\n")
+            return
+        elif sample.successful.stdout is None:
+            self.logger.critical(
+                ("dnsperf ran successfully, but did not get output on stdout.\n" f"Got results: {sample}\n")
+            )
+            return
+
+        stdout: DnsperfStdout = self.parse_stdout(sample.successful.stdout)
+        cfg: DnsperfConfig = DnsperfConfig.new(stdout, self.config, load=load)
+        self.logger.info(stdout)
+        self.logger.info(cfg)
+
+        return self.create_new_result(
+            data=dataclasses.asdict(stdout), config=dataclasses.asdict(cfg), tag="results"
+        )
+
+    def collect(self) -> Iterable[BenchmarkResult]:
+        """Run the dnsperf benchmark and collect results."""
+
+        self.logger.info("Starting dnsperf")
+        first_result: BenchmarkResult = self._one_trial()
+        yield first_result
+
+        max_qps = first_result.data["throughput"]
+        allowed_loads = (
+            np.linspace(
```

Comment:
Would it be possible to do this without numpy in order to save a dependency? The built-in ``range`` function includes a step parameter.